### DISCUSSION TITLE
Logging strategies on iOS

### DISCUSSION THREAD

**Discussion by emanuelef** [MobileNativeFoundation/discussions#104] [General]
Hi,
Just curious to know what others are doing nowadays about Logging, especially in iOS.
1. Are you using directly Apple unified logging system or a library (CocoaLumberjack, SwiftyBeaver...) ?
2. Do you log on files on the devices as well ? In case did you add a feature to send the logs in your app on some occasions, like after crashing ?
3. Do you upload logs to some cloud service ?
4. How much do you log in general and how did it help with some issues on production ?
5. Do you also have logs to understand more about users behaviours ?

Thanks

**Comment by emanuelef** [comment]
I'll start by answering with what we are doing at the moment:

1. We were using CocoaLumberjack but moving to Apple Unified Logging (AUL). One thing to consider is that log levels are different than other logging systems.
2. We used to save file for the last hours, but AUL should take care of that. How to extract and use them from a end user device doesn't seem to be easy though.
3. Not at the moment, but we are using Crashlytics and save some custom logs there to give more context in case of crashes.
4. Mostly log errors and state changes
5. Not at the moment

**Comment by dive** [comment]
I believe this is the most up-to-date overview of the matter – [Logging in Swift](https://steipete.me/posts/logging-in-swift/) by @steipete. `OSLogStore` is the most problematic part for many applications. We need logs in half of the cases, and it is not feasible to rely on a solution without the option to upload/parse the logs on the device. `OSLogStore` looks more or less fine, but it is iOS 14 and up. Have to wait for a few more years.
[CocoaLumberjack](https://github.com/CocoaLumberjack/CocoaLumberjack/) is the way to go for now.
